crates/btcio/src/broadcaster/task.rs

```diff
@@ -35,147 +35,144 @@ pub async fn broadcaster_task(
             Some((idx, txentry)) = entry_receiver.recv() => {
                 let txid: Option<Txid> = ops.get_txid_async(idx).await?.map(Into::into);
                 info!(%idx, ?txid, "Received txentry");
-
-                // Insert into state's unfinalized entries. Need not update next_idx because that
-                // will be handled in state.next() call
-                state.unfinalized_entries.insert(idx, txentry);
+                state.unfinalized_entries.push((idx, txentry));
             }
         }
 
-        let (updated_entries, to_remove) = process_unfinalized_entries(
+        // Process any unfinalized entries
+        let updated_entries = process_unfinalized_entries(
             &state.unfinalized_entries,
             ops.clone(),
             rpc_client.as_ref(),
             params.as_ref(),
         )
         .await
-        .map_err(|e| {
+        .inspect_err(|e| {
             error!(%e, "broadcaster exiting");
-            e
         })?;
 
-        for idx in to_remove {
-            _ = state.unfinalized_entries.remove(&idx);
+        // Update in db
+        for (idx, txentry) in updated_entries.iter() {
+            ops.put_tx_entry_by_idx_async(*idx, txentry.clone()).await?;
         }
 
-        state.next(updated_entries, &ops).await?;
+        // Update the state.
+        state.update(updated_entries, &ops).await?;
     }
 }
 
-/// Processes unfinalized entries and returns entries idxs that are finalized
+/// Processes unfinalized entries and returns entries idxs that are updated.
 async fn process_unfinalized_entries(
-    unfinalized_entries: &BTreeMap<u64, L1TxEntry>,
+    unfinalized_entries: &[(u64, L1TxEntry)],
```

Comment:
Could this be a `impl Iterator<Item = &(u64, L1TxEntry)>`?  We immediately call `.iter()` on it anyways.